Date: Tue, 22 Apr 2025 20:32:07 +0000
Subject: AL2023 6.12: Update lustrefsx to 2.15.6-fsx18 commit


diff --git a/fs/lustre/config.h b/fs/lustre/config.h
--- a/fs/lustre/config.h
+++ b/fs/lustre/config.h
@@ -1289,7 +1289,7 @@
 #define LUSTRE_PATCH 6
 
 /* A copy of PACKAGE_VERSION */
-#define LUSTRE_VERSION_STRING "2.15.6_590_g46ce8c1_dirty"
+#define LUSTRE_VERSION_STRING "2.15.6_599_gd5769bb_dirty"
 
 /* maximum number of MDS threads */
 /* #undef MDS_MAX_THREADS */
@@ -1319,7 +1319,7 @@
 #define PACKAGE_NAME "Lustre"
 
 /* Define to the full name and version of this package. */
-#define PACKAGE_STRING "Lustre 2.15.6_590_g46ce8c1_dirty"
+#define PACKAGE_STRING "Lustre 2.15.6_599_gd5769bb_dirty"
 
 /* Define to the one symbol short name of this package. */
 #define PACKAGE_TARNAME "lustre"
@@ -1328,7 +1328,7 @@
 #define PACKAGE_URL ""
 
 /* Define to the version of this package. */
-#define PACKAGE_VERSION "2.15.6_590_g46ce8c1_dirty"
+#define PACKAGE_VERSION "2.15.6_599_gd5769bb_dirty"
 
 /* name of parallel fsck program */
 #define PFSCK "fsck"
@@ -1372,7 +1372,7 @@
 /* #undef USE_HEALTH_CHECK_WRITE */
 
 /* Version number of package */
-#define VERSION "2.15.6_590_g46ce8c1_dirty"
+#define VERSION "2.15.6_599_gd5769bb_dirty"
 
 /* vfs_setxattr() value argument is non-const */
 #define VFS_SETXATTR_VALUE(value) ((void *)(value))
diff --git a/fs/lustre/include/lustre_dlm.h b/fs/lustre/include/lustre_dlm.h
--- a/fs/lustre/include/lustre_dlm.h
+++ b/fs/lustre/include/lustre_dlm.h
@@ -61,7 +61,7 @@ extern struct kset *ldlm_svc_kset;
 #define OBD_LDLM_DEVICENAME  "ldlm"
 
 #define LDLM_DEFAULT_LRU_SIZE (100 * num_online_cpus())
-#define LDLM_DEFAULT_MAX_ALIVE		3900	/* 3900 seconds ~65 min */
+#define LDLM_DEFAULT_LRU_MAX_AGE	600	/* 600 seconds = 10 min */
 #define LDLM_CTIME_AGE_LIMIT (10)
 /* if client lock is unused for that time it can be cancelled if any other
  * client shows interest in that lock, e.g. glimpse is occured. */
diff --git a/fs/lustre/include/lustre_idmap.h b/fs/lustre/include/lustre_idmap.h
--- a/fs/lustre/include/lustre_idmap.h
+++ b/fs/lustre/include/lustre_idmap.h
@@ -62,6 +62,7 @@
 struct lu_ucred;
 
 extern void lustre_groups_from_list(struct group_info *ginfo, gid_t *glist);
+extern void lustre_list_from_groups(gid_t *glist, struct group_info *ginfo);
 extern void lustre_groups_sort(struct group_info *group_info);
 extern int lustre_in_group_p(struct lu_ucred *mu, gid_t grp);
 
diff --git a/fs/lustre/ldlm/ldlm_reclaim.c b/fs/lustre/ldlm/ldlm_reclaim.c
--- a/fs/lustre/ldlm/ldlm_reclaim.c
+++ b/fs/lustre/ldlm/ldlm_reclaim.c
@@ -227,7 +227,7 @@ static void ldlm_reclaim_res(struct ldlm_namespace *ns, int *count,
 
 #define LDLM_RECLAIM_BATCH	512
 #define LDLM_RECLAIM_AGE_MIN	(300 * NSEC_PER_SEC)
-#define LDLM_RECLAIM_AGE_MAX	(LDLM_DEFAULT_MAX_ALIVE * NSEC_PER_SEC * 3 / 4)
+#define LDLM_RECLAIM_AGE_MAX	(LDLM_DEFAULT_LRU_MAX_AGE * NSEC_PER_SEC * 3 / 4)
 
 static inline s64 ldlm_reclaim_age(void)
 {
diff --git a/fs/lustre/ldlm/ldlm_resource.c b/fs/lustre/ldlm/ldlm_resource.c
--- a/fs/lustre/ldlm/ldlm_resource.c
+++ b/fs/lustre/ldlm/ldlm_resource.c
@@ -967,7 +967,7 @@ struct ldlm_namespace *ldlm_namespace_new(struct obd_device *obd, char *name,
 	ns->ns_max_unused         = LDLM_DEFAULT_LRU_SIZE;
 	ns->ns_cancel_batch       = LDLM_DEFAULT_LRU_SHRINK_BATCH;
 	ns->ns_recalc_pct         = LDLM_DEFAULT_SLV_RECALC_PCT;
-	ns->ns_max_age            = ktime_set(LDLM_DEFAULT_MAX_ALIVE, 0);
+	ns->ns_max_age            = ktime_set(LDLM_DEFAULT_LRU_MAX_AGE, 0);
 	ns->ns_ctime_age_limit    = LDLM_CTIME_AGE_LIMIT;
 	ns->ns_dirty_age_limit    = ktime_set(LDLM_DIRTY_AGE_LIMIT, 0);
 	ns->ns_timeouts           = 0;
diff --git a/fs/lustre/llite/llite_internal.h b/fs/lustre/llite/llite_internal.h
--- a/fs/lustre/llite/llite_internal.h
+++ b/fs/lustre/llite/llite_internal.h
@@ -1163,6 +1163,10 @@ int ll_md_blocking_ast(struct ldlm_lock *, struct ldlm_lock_desc *,
 struct dentry *ll_splice_alias(struct inode *inode, struct dentry *de);
 int ll_rmdir_entry(struct inode *dir, char *name, int namelen);
 void ll_update_times(struct ptlrpc_request *request, struct inode *inode);
+int ll_intent_lock(struct obd_export *exp, struct md_op_data *op_data,
+		   struct lookup_intent *it, struct ptlrpc_request **reqp,
+		   ldlm_blocking_callback cb_blocking, __u64 extra_lock_flags,
+		   bool tryagain);
 
 /* llite/rw.c */
 int ll_writepage(struct page *page, struct writeback_control *wbc);
diff --git a/fs/lustre/llite/namei.c b/fs/lustre/llite/namei.c
--- a/fs/lustre/llite/namei.c
+++ b/fs/lustre/llite/namei.c
@@ -805,6 +805,236 @@ static int ll_lookup_it_finish(struct ptlrpc_request *request,
 	return rc;
 }
 
+static int get_acl_from_req(struct ptlrpc_request *req, struct posix_acl **acl)
+{
+	struct mdt_body	*body;
+	void *buf;
+	int rc;
+
+	body = req_capsule_server_get(&req->rq_pill, &RMF_MDT_BODY);
+	if (!body->mbo_aclsize) {
+		*acl = NULL;
+		return 0;
+	}
+
+	buf = req_capsule_server_sized_get(&req->rq_pill, &RMF_ACL,
+					   body->mbo_aclsize);
+	if (!buf)
+		return -EPROTO;
+
+	*acl = posix_acl_from_xattr(&init_user_ns, buf, body->mbo_aclsize);
+	if (IS_ERR_OR_NULL(*acl)) {
+		rc = *acl ? PTR_ERR(*acl) : 0;
+		CDEBUG(D_SEC, "convert xattr to acl: %d\n", rc);
+		return rc;
+	}
+
+	rc = posix_acl_valid(&init_user_ns, *acl);
+	if (rc) {
+		CDEBUG(D_SEC, "validate acl: %d\n", rc);
+		posix_acl_release(*acl);
+		return rc;
+	}
+
+	return 0;
+}
+
+static inline int accmode_from_openflags(u64 open_flags)
+{
+	unsigned int may_mask = 0;
+
+	if (open_flags & (FMODE_READ | FMODE_PREAD))
+		may_mask |= MAY_READ;
+	if (open_flags & (FMODE_WRITE | FMODE_PWRITE))
+		may_mask |= MAY_WRITE;
+	if (open_flags & FMODE_EXEC)
+		may_mask = MAY_EXEC;
+
+	return may_mask;
+}
+
+static __u32 get_uc_group_from_acl(const struct posix_acl *acl, int want)
+{
+	const struct posix_acl_entry *pa, *pe;
+
+	FOREACH_ACL_ENTRY(pa, acl, pe) {
+		switch (pa->e_tag) {
+		case ACL_GROUP_OBJ:
+		case ACL_GROUP:
+			if (in_group_p(pa->e_gid) &&
+			    (pa->e_perm & want) == want)
+				return (__u32)from_kgid(&init_user_ns,
+							pa->e_gid);
+			break;
+		default:
+			/* nothing to do */
+			break;
+		}
+	}
+
+	return (__u32)__kgid_val(INVALID_GID);
+}
+
+struct ll_grp_it {
+	struct group_info *gi;
+	int pos;
+};
+
+static struct ll_grp_it *ll_group_iter_start(void)
+{
+	struct ll_grp_it *git;
+
+	OBD_ALLOC_PTR(git);
+	if (!git)
+		return NULL;
+
+	git->gi = get_current_groups();
+	git->pos = 0;
+
+	return git;
+}
+
+static u32 ll_group_iter_next(struct ll_grp_it *git)
+{
+	int idx;
+
+	LASSERT(git);
+	idx = git->pos;
+	git->pos++;
+
+	/* -1 is treated as invalid */
+	if (idx >= git->gi->ngroups || idx >= NGROUPS_MAX)
+		return -1;
+
+	/* TODO @timday: This iterator doesn't work on
+	 * CentOS, so give up...
+	 */
+#ifdef HAVE_GROUP_INFO_GID
+	return git->gi->gid[idx].val;
+#else /* !HAVE_GROUP_INFO_GID */
+	return -1;
+#endif /* HAVE_GROUP_INFO_GID */
+}
+
+static bool ll_group_iter_is_done(struct ll_grp_it *git)
+{
+	LASSERT(git);
+
+	if (git->pos >= git->gi->ngroups || git->pos >= NGROUPS_MAX)
+		return true;
+
+	return false;
+}
+
+static void ll_group_iter_stop(struct ll_grp_it *git)
+{
+	LASSERT(git);
+	put_group_info(git->gi);
+	OBD_FREE_PTR(git);
+}
+
+/* This function implements a retry mechanism on top of md_intent_lock().
+ * This is useful because the client can provide at most 2 supplementary
+ * groups in the request sent to the MDS, but sometimes it does not know
+ * which ones are useful for credentials calculation on server side. For
+ * instance in case of lookup, the client does not have the child inode yet
+ * when it sends the intent lock request.
+ *
+ * Hopefully, the server can hint at the useful groups, by putting in the
+ * request reply the target inode's GID, and also its ACL.
+ * So in case the server replies -EACCES, we check the user's credentials
+ * against those, and try again the intent lock request if we find a matching
+ * supplementary group.
+ *
+ * If that still doesn't work, we can cycle through the available
+ * supplementary groups rather than giving up early.
+ */
+int ll_intent_lock(struct obd_export *exp, struct md_op_data *op_data,
+		   struct lookup_intent *it, struct ptlrpc_request **reqp,
+		   ldlm_blocking_callback cb_blocking, __u64 extra_lock_flags,
+		   bool tryagain)
+{
+	struct ll_grp_it *git = ll_group_iter_start();
+	struct ptlrpc_request *req = *reqp;
+	bool more_viable_gids = false;
+	int rc = 0;
+
+	ENTRY;
+
+again:
+	if (op_data->op_suppgids[0] == -1)
+		op_data->op_suppgids[0] = ll_group_iter_next(git);
+
+	if (op_data->op_suppgids[1] == -1)
+		op_data->op_suppgids[1] = ll_group_iter_next(git);
+
+	rc = md_intent_lock(exp, op_data, it, reqp, cb_blocking,
+			    extra_lock_flags);
+
+	CDEBUG(D_VFSTRACE,
+	       "intent lock %d on i1 "DFID" suppgids[%d][%d] rc[%d] tryagain[%i] iter[%i]\n",
+	       it->it_op, PFID(&op_data->op_fid1),
+	       op_data->op_suppgids[0], op_data->op_suppgids[1], rc,
+	       tryagain, ll_group_iter_is_done(git));
+
+	/* We can't (or won't) try again! */
+	if (rc != -EACCES || !tryagain)
+		GOTO(out, rc);
+
+	/* These GIDs don't work */
+	op_data->op_suppgids[0] = -1;
+	op_data->op_suppgids[1] = -1;
+
+	/* If the MDS allows the client to chgrp (CFS_SETGRP_PERM), but the
+	 * client does not know which suppgid should be sent to the MDS, or
+	 * some other(s) changed the target file's GID after this RPC sent
+	 * to the MDS with the suppgid as the original GID, then we should
+	 * try again with right suppgid.
+	 */
+	if (req && it->it_op & IT_OPEN &&
+	    it_disposition(it, DISP_OPEN_DENY)) {
+		int accmode = accmode_from_openflags(it->it_flags);
+		struct posix_acl *acl;
+		struct mdt_body *body;
+
+		LASSERT(req);
+
+		rc = get_acl_from_req(*reqp, &acl);
+		if (rc || !acl) {
+			op_data->op_suppgids[0] = -1;
+		} else {
+			op_data->op_suppgids[0] = get_uc_group_from_acl(acl, accmode);
+			posix_acl_release(acl);
+		}
+
+		body = req_capsule_server_get(&req->rq_pill, &RMF_MDT_BODY);
+
+		if (in_group_p(make_kgid(&init_user_ns, body->mbo_gid)))
+			op_data->op_suppgids[1] = body->mbo_gid;
+	}
+
+	more_viable_gids = !ll_group_iter_is_done(git) ||
+	  op_data->op_suppgids[0] != -1 ||
+	  op_data->op_suppgids[1] != -1;
+
+	/* Keep trying! */
+	if (more_viable_gids) {
+		if (!(it->it_flags & MDS_OPEN_BY_FID))
+			fid_zero(&op_data->op_fid2);
+
+		if (req)
+			ptlrpc_req_finished(req);
+
+		req = NULL;
+		ll_intent_release(it);
+		GOTO(again, rc = 0);
+	}
+
+out:
+	ll_group_iter_stop(git);
+	RETURN(rc);
+}
+
 static struct dentry *ll_lookup_it(struct inode *parent, struct dentry *dentry,
 				   struct lookup_intent *it,
 				   void **secctx, __u32 *secctxlen,
@@ -1014,34 +1244,9 @@ static struct dentry *ll_lookup_it(struct inode *parent, struct dentry *dentry,
 		it->it_flags |= MDS_OPEN_PCC;
 	}
 
-	rc = md_intent_lock(ll_i2mdexp(parent), op_data, it, &req,
-			    &ll_md_blocking_ast, 0);
-	/* If the MDS allows the client to chgrp (CFS_SETGRP_PERM), but the
-	 * client does not know which suppgid should be sent to the MDS, or
-	 * some other(s) changed the target file's GID after this RPC sent
-	 * to the MDS with the suppgid as the original GID, then we should
-	 * try again with right suppgid. */
-	if (rc == -EACCES && it->it_op & IT_OPEN &&
-	    it_disposition(it, DISP_OPEN_DENY)) {
-		struct mdt_body *body;
-
-		LASSERT(req != NULL);
-
-		body = req_capsule_server_get(&req->rq_pill, &RMF_MDT_BODY);
-		if (op_data->op_suppgids[0] == body->mbo_gid ||
-		    op_data->op_suppgids[1] == body->mbo_gid ||
-		    !in_group_p(make_kgid(&init_user_ns, body->mbo_gid)))
-			GOTO(out, retval = ERR_PTR(-EACCES));
-
-		fid_zero(&op_data->op_fid2);
-		op_data->op_suppgids[1] = body->mbo_gid;
-		ptlrpc_req_finished(req);
-		req = NULL;
-		ll_intent_release(it);
-		rc = md_intent_lock(ll_i2mdexp(parent), op_data, it, &req,
-				    &ll_md_blocking_ast, 0);
-	}
-
+	/* Try an intent lock with more GIDs */
+	rc = ll_intent_lock(ll_i2mdexp(parent), op_data, it, &req,
+			    &ll_md_blocking_ast, 0, true);
 	if (rc < 0)
 		GOTO(out, retval = ERR_PTR(rc));
 
diff --git a/fs/lustre/lmv/lmv_intent.c b/fs/lustre/lmv/lmv_intent.c
--- a/fs/lustre/lmv/lmv_intent.c
+++ b/fs/lustre/lmv/lmv_intent.c
@@ -49,22 +49,43 @@
 #include <lprocfs_status.h>
 #include "lmv_internal.h"
 
-static int lmv_intent_remote(struct obd_export *exp, struct lookup_intent *it,
+/* TODO @timday: Is this the only op that needs op_data_tmp? */
+static struct md_op_data *lmv_remote_prep_md_op(struct md_op_data *op_data)
+{
+	struct md_op_data *op_data_tmp;
+	OBD_ALLOC_PTR(op_data_tmp);
+	if (!op_data_tmp)
+		return NULL;
+
+	op_data_tmp->op_suppgids[0] = op_data->op_suppgids[0];
+	op_data_tmp->op_suppgids[1] = op_data->op_suppgids[1];
+
+	return op_data_tmp;
+}
+
+static void lmv_remote_finish_md_op(struct md_op_data *op_data_tmp)
+{
+	OBD_FREE_PTR(op_data_tmp);
+}
+
+static int lmv_intent_remote(struct obd_export *exp, struct md_op_data *op_data,
+			     struct lookup_intent *it,
 			     const struct lu_fid *parent_fid,
 			     struct ptlrpc_request **reqp,
 			     ldlm_blocking_callback cb_blocking,
 			     __u64 extra_lock_flags,
 			     const char *secctx_name, __u32 secctx_name_size)
 {
-	struct obd_device	*obd = exp->exp_obd;
-	struct lmv_obd		*lmv = &obd->u.lmv;
-	struct ptlrpc_request	*req = NULL;
-	struct lustre_handle	plock;
-	struct md_op_data	*op_data;
-	struct lmv_tgt_desc	*tgt;
-	struct mdt_body		*body;
-	int			pmode;
-	int			rc = 0;
+	struct obd_device *obd = exp->exp_obd;
+	struct lmv_obd *lmv = &obd->u.lmv;
+	struct ptlrpc_request *req = NULL;
+	struct md_op_data *op_data_tmp;
+	struct lustre_handle plock;
+	struct lmv_tgt_desc *tgt;
+	struct mdt_body *body;
+	int rc = 0;
+	int pmode;
+
 	ENTRY;
 
 	body = req_capsule_server_get(&(*reqp)->rq_pill, &RMF_MDT_BODY);
@@ -89,36 +110,36 @@ static int lmv_intent_remote(struct obd_export *exp, struct lookup_intent *it,
 	if (IS_ERR(tgt))
 		GOTO(out, rc = PTR_ERR(tgt));
 
-	OBD_ALLOC_PTR(op_data);
-	if (op_data == NULL)
+	op_data_tmp = lmv_remote_prep_md_op(op_data);
+	if (!op_data_tmp)
 		GOTO(out, rc = -ENOMEM);
 
-	op_data->op_fid1 = body->mbo_fid1;
+	op_data_tmp->op_fid1 = body->mbo_fid1;
 	/* Sent the parent FID to the remote MDT */
 	if (parent_fid != NULL) {
 		/* The parent fid is only for remote open to
 		 * check whether the open is from OBF,
 		 * see mdt_cross_open */
 		LASSERT(it->it_op & IT_OPEN);
-		op_data->op_fid2 = *parent_fid;
+		op_data_tmp->op_fid2 = *parent_fid;
 	}
 
-	op_data->op_bias = MDS_CROSS_REF;
-	op_data->op_cli_flags = CLI_NO_SLOT;
+	op_data_tmp->op_bias = MDS_CROSS_REF;
+	op_data_tmp->op_cli_flags = CLI_NO_SLOT;
 	CDEBUG(D_INODE, "REMOTE_INTENT with fid="DFID" -> mds #%u\n",
 	       PFID(&body->mbo_fid1), tgt->ltd_index);
 
 	/* ask for security context upon intent */
 	if (it->it_op & (IT_LOOKUP | IT_GETATTR | IT_OPEN) &&
 	    secctx_name_size != 0 && secctx_name != NULL) {
-		op_data->op_file_secctx_name = secctx_name;
-		op_data->op_file_secctx_name_size = secctx_name_size;
+		op_data_tmp->op_file_secctx_name = secctx_name;
+		op_data_tmp->op_file_secctx_name_size = secctx_name_size;
 		CDEBUG(D_SEC, "'%.*s' is security xattr to fetch for "
 		       DFID"\n",
 		       secctx_name_size, secctx_name, PFID(&body->mbo_fid1));
 	}
 
-	rc = md_intent_lock(tgt->ltd_exp, op_data, it, &req, cb_blocking,
+	rc = md_intent_lock(tgt->ltd_exp, op_data_tmp, it, &req, cb_blocking,
 			    extra_lock_flags);
         if (rc)
                 GOTO(out_free_op_data, rc);
@@ -139,16 +160,15 @@ static int lmv_intent_remote(struct obd_export *exp, struct lookup_intent *it,
 		it->it_lock_mode = pmode;
 	}
 
-	EXIT;
 out_free_op_data:
-	OBD_FREE_PTR(op_data);
+	lmv_remote_finish_md_op(op_data_tmp);
 out:
 	if (rc && pmode)
 		ldlm_lock_decref(&plock, pmode);
 
 	ptlrpc_req_finished(*reqp);
 	*reqp = req;
-	return rc;
+	RETURN(rc);
 }
 
 int lmv_revalidate_slaves(struct obd_export *exp,
@@ -406,7 +426,7 @@ static int lmv_intent_open(struct obd_export *exp, struct md_op_data *op_data,
 
 	/* Not cross-ref case, just get out of here. */
 	if (unlikely((body->mbo_valid & OBD_MD_MDS))) {
-		rc = lmv_intent_remote(exp, it, &op_data->op_fid1, reqp,
+		rc = lmv_intent_remote(exp, op_data, it, &op_data->op_fid1, reqp,
 				       cb_blocking, extra_lock_flags,
 				       op_data->op_file_secctx_name,
 				       op_data->op_file_secctx_name_size);
@@ -530,12 +550,13 @@ lmv_intent_lookup(struct obd_export *exp, struct md_op_data *op_data,
 
 	/* Not cross-ref case, just get out of here. */
 	if (unlikely((body->mbo_valid & OBD_MD_MDS))) {
-		rc = lmv_intent_remote(exp, it, NULL, reqp, cb_blocking,
+		rc = lmv_intent_remote(exp, op_data, it, NULL, reqp, cb_blocking,
 				       extra_lock_flags,
 				       op_data->op_file_secctx_name,
 				       op_data->op_file_secctx_name_size);
-		if (rc != 0)
+		if (rc)
 			RETURN(rc);
+
 		body = req_capsule_server_get(&(*reqp)->rq_pill, &RMF_MDT_BODY);
 		if (body == NULL)
 			RETURN(-EPROTO);
diff --git a/fs/lustre/ptlrpc/connection.c b/fs/lustre/ptlrpc/connection.c
--- a/fs/lustre/ptlrpc/connection.c
+++ b/fs/lustre/ptlrpc/connection.c
@@ -84,6 +84,7 @@ ptlrpc_connection_get(struct lnet_process_id peer4, lnet_nid_t self,
 	ENTRY;
 
 	peer4.nid = LNetPrimaryNID(peer4.nid);
+	CDEBUG(D_INFO, "Found correct NID -> %s\n", libcfs_nid2str(peer4.nid));
 	lnet_pid4_to_pid(peer4, &peer);
 	conn = rhashtable_lookup_fast(&conn_hash, &peer, conn_hash_params);
 	if (conn) {
@@ -98,8 +99,11 @@ ptlrpc_connection_get(struct lnet_process_id peer4, lnet_nid_t self,
 	conn->c_peer = peer;
 	lnet_nid4_to_nid(self, &conn->c_self);
 	atomic_set(&conn->c_refcount, 1);
-	if (uuid)
-		obd_str2uuid(&conn->c_remote_uuid, uuid->uuid);
+	if (uuid) {
+		obd_str2uuid(&conn->c_remote_uuid, libcfs_nid2str(peer4.nid));
+		obd_str2uuid(uuid, libcfs_nid2str(peer4.nid));
+		CDEBUG(D_INFO, "New UUID -> %s\n", uuid->uuid);
+	}
 
 	/*
 	 * Add the newly created conn to the hash, on key collision we
diff --git a/fs/lustre/ptlrpc/events.c b/fs/lustre/ptlrpc/events.c
--- a/fs/lustre/ptlrpc/events.c
+++ b/fs/lustre/ptlrpc/events.c
@@ -537,6 +537,7 @@ int ptlrpc_uuid_to_peer(struct obd_uuid *uuid,
 	__u32 order;
 	lnet_nid_t dst_nid;
 	lnet_nid_t src_nid;
+	lnet_nid_t real_nid;
 
 	peer->pid = LNET_PID_LUSTRE;
 
@@ -568,6 +569,13 @@ int ptlrpc_uuid_to_peer(struct obd_uuid *uuid,
 		}
 	}
 
+	real_nid = LNetPrimaryNID(peer->nid);
+	CDEBUG(D_INFO, "Found correct NID -> %s\n", libcfs_nid2str(real_nid));
+	if (uuid) {
+		obd_str2uuid(uuid, libcfs_nid2str(real_nid));
+		CDEBUG(D_INFO, "New UUID -> %s\n", uuid->uuid);
+	}
+
 	CDEBUG(D_NET, "%s->%s\n", uuid->uuid, libcfs_id2str(*peer));
 	return rc;
 }
diff --git a/net/lnet/libcfs/libcfs/util/libcfs_la-l_ioctl.lo b/net/lnet/libcfs/libcfs/util/libcfs_la-l_ioctl.lo
deleted file mode 100644
--- a/net/lnet/libcfs/libcfs/util/libcfs_la-l_ioctl.lo
+++ /dev/null
@@ -1,12 +0,0 @@
-# util/libcfs_la-l_ioctl.lo - a libtool object file
-# Generated by libtool (GNU libtool) 2.4.7
-#
-# Please DO NOT delete this file!
-# It is necessary for linking the library.
-
-# Name of the PIC object.
-pic_object='.libs/libcfs_la-l_ioctl.o'
-
-# Name of the non-PIC object
-non_pic_object='libcfs_la-l_ioctl.o'
-
diff --git a/net/lnet/libcfs/libcfs/util/libcfs_la-nidstrings.lo b/net/lnet/libcfs/libcfs/util/libcfs_la-nidstrings.lo
deleted file mode 100644
--- a/net/lnet/libcfs/libcfs/util/libcfs_la-nidstrings.lo
+++ /dev/null
@@ -1,12 +0,0 @@
-# util/libcfs_la-nidstrings.lo - a libtool object file
-# Generated by libtool (GNU libtool) 2.4.7
-#
-# Please DO NOT delete this file!
-# It is necessary for linking the library.
-
-# Name of the PIC object.
-pic_object='.libs/libcfs_la-nidstrings.o'
-
-# Name of the non-PIC object
-non_pic_object='libcfs_la-nidstrings.o'
-
diff --git a/net/lnet/libcfs/libcfs/util/libcfs_la-param.lo b/net/lnet/libcfs/libcfs/util/libcfs_la-param.lo
deleted file mode 100644
--- a/net/lnet/libcfs/libcfs/util/libcfs_la-param.lo
+++ /dev/null
@@ -1,12 +0,0 @@
-# util/libcfs_la-param.lo - a libtool object file
-# Generated by libtool (GNU libtool) 2.4.7
-#
-# Please DO NOT delete this file!
-# It is necessary for linking the library.
-
-# Name of the PIC object.
-pic_object='.libs/libcfs_la-param.o'
-
-# Name of the non-PIC object
-non_pic_object='libcfs_la-param.o'
-
diff --git a/net/lnet/libcfs/libcfs/util/libcfs_la-parser.lo b/net/lnet/libcfs/libcfs/util/libcfs_la-parser.lo
deleted file mode 100644
--- a/net/lnet/libcfs/libcfs/util/libcfs_la-parser.lo
+++ /dev/null
@@ -1,12 +0,0 @@
-# util/libcfs_la-parser.lo - a libtool object file
-# Generated by libtool (GNU libtool) 2.4.7
-#
-# Please DO NOT delete this file!
-# It is necessary for linking the library.
-
-# Name of the PIC object.
-pic_object='.libs/libcfs_la-parser.o'
-
-# Name of the non-PIC object
-non_pic_object='libcfs_la-parser.o'
-
diff --git a/net/lnet/libcfs/libcfs/util/libcfs_la-string.lo b/net/lnet/libcfs/libcfs/util/libcfs_la-string.lo
deleted file mode 100644
--- a/net/lnet/libcfs/libcfs/util/libcfs_la-string.lo
+++ /dev/null
@@ -1,12 +0,0 @@
-# util/libcfs_la-string.lo - a libtool object file
-# Generated by libtool (GNU libtool) 2.4.7
-#
-# Please DO NOT delete this file!
-# It is necessary for linking the library.
-
-# Name of the PIC object.
-pic_object='.libs/libcfs_la-string.o'
-
-# Name of the non-PIC object
-non_pic_object='libcfs_la-string.o'
-
