Date: Wed, 13 Dec 2023 02:51:38 +0000
Subject: Revert "objtool: Propagate early errors"

This reverts commit 21f99a5adbc522c8e8126132e94d353297fd3c16.

diff --git a/tools/objtool/objtool.c b/tools/objtool/objtool.c
--- a/tools/objtool/objtool.c
+++ b/tools/objtool/objtool.c
@@ -146,5 +146,7 @@ int main(int argc, const char **argv)
 	exec_cmd_init("objtool", UNUSED, UNUSED, UNUSED);
 	pager_init(UNUSED);
 
-	return objtool_run(argc, argv);
+	objtool_run(argc, argv);
+
+	return 0;
 }
