Date: Fri, 12 Jul 2024 00:09:15 +0000
Subject: AL2023-6.1-Update-ena-driver-to-2.12.3g

Signed-off-by: Arthur Kiyanovski <akiyano@amazon.com>

diff --git a/drivers/amazon/net/ena/Makefile b/drivers/amazon/net/ena/Makefile
--- a/drivers/amazon/net/ena/Makefile
+++ b/drivers/amazon/net/ena/Makefile
@@ -1,7 +1,7 @@
 #
 # Makefile for the Elastic Network Adapter (ENA) device drivers.
 # ENA Source is: https://github.com/amzn/amzn-drivers.
-# Current ENA source is based on ena_linux_2.12.0 tag.
+# Current ENA source is based on ena_linux_2.12.3 tag.
 #
 
 obj-$(CONFIG_AMAZON_ENA_ETHERNET) += ena.o
diff --git a/drivers/amazon/net/ena/config.h b/drivers/amazon/net/ena/config.h
--- a/drivers/amazon/net/ena/config.h
+++ b/drivers/amazon/net/ena/config.h
@@ -4,4 +4,5 @@
 #define ENA_HAVE_XDP_DO_FLUSH 1
 #define ENA_HAVE_CPUMASK_LOCAL_SPREAD 1
 #define ENA_HAVE_UPDATE_AFFINITY_HINT 1
+#define ENA_HAVE_ETH_HW_ADDR_SET 1
 #endif /* _ENA_CONFIG_H_ */
diff --git a/drivers/amazon/net/ena/ena_com.c b/drivers/amazon/net/ena/ena_com.c
--- a/drivers/amazon/net/ena/ena_com.c
+++ b/drivers/amazon/net/ena/ena_com.c
@@ -321,7 +321,6 @@ static int ena_com_init_io_sq(struct ena_com_dev *ena_dev,
 			      struct ena_com_io_sq *io_sq)
 {
 	size_t size;
-	int dev_node = 0;
 
 	memset(&io_sq->desc_addr, 0x0, sizeof(io_sq->desc_addr));
 
@@ -335,12 +334,9 @@ static int ena_com_init_io_sq(struct ena_com_dev *ena_dev,
 	io_sq->bus = ena_dev->bus;
 
 	if (io_sq->mem_queue_type == ENA_ADMIN_PLACEMENT_POLICY_HOST) {
-		dev_node = dev_to_node(ena_dev->dmadev);
-		set_dev_node(ena_dev->dmadev, ctx->numa_node);
 		io_sq->desc_addr.virt_addr =
 			dma_zalloc_coherent(ena_dev->dmadev, size, &io_sq->desc_addr.phys_addr,
 					    GFP_KERNEL);
-		set_dev_node(ena_dev->dmadev, dev_node);
 		if (!io_sq->desc_addr.virt_addr) {
 			io_sq->desc_addr.virt_addr =
 				dma_zalloc_coherent(ena_dev->dmadev, size,
@@ -364,10 +360,7 @@ static int ena_com_init_io_sq(struct ena_com_dev *ena_dev,
 		size = (size_t)io_sq->bounce_buf_ctrl.buffer_size *
 			io_sq->bounce_buf_ctrl.buffers_num;
 
-		dev_node = dev_to_node(ena_dev->dmadev);
-		set_dev_node(ena_dev->dmadev, ctx->numa_node);
 		io_sq->bounce_buf_ctrl.base_buffer = devm_kzalloc(ena_dev->dmadev, size, GFP_KERNEL);
-		set_dev_node(ena_dev->dmadev, dev_node);
 		if (!io_sq->bounce_buf_ctrl.base_buffer)
 			io_sq->bounce_buf_ctrl.base_buffer =
 				devm_kzalloc(ena_dev->dmadev, size, GFP_KERNEL);
@@ -407,7 +400,6 @@ static int ena_com_init_io_cq(struct ena_com_dev *ena_dev,
 			      struct ena_com_io_cq *io_cq)
 {
 	size_t size;
-	int prev_node = 0;
 
 	memset(&io_cq->cdesc_addr, 0x0, sizeof(io_cq->cdesc_addr));
 
@@ -420,11 +412,8 @@ static int ena_com_init_io_cq(struct ena_com_dev *ena_dev,
 	size = io_cq->cdesc_entry_size_in_bytes * io_cq->q_depth;
 	io_cq->bus = ena_dev->bus;
 
-	prev_node = dev_to_node(ena_dev->dmadev);
-	set_dev_node(ena_dev->dmadev, ctx->numa_node);
 	io_cq->cdesc_addr.virt_addr =
 		dma_zalloc_coherent(ena_dev->dmadev, size, &io_cq->cdesc_addr.phys_addr, GFP_KERNEL);
-	set_dev_node(ena_dev->dmadev, prev_node);
 	if (!io_cq->cdesc_addr.virt_addr) {
 		io_cq->cdesc_addr.virt_addr =
 			dma_zalloc_coherent(ena_dev->dmadev, size, &io_cq->cdesc_addr.phys_addr,
diff --git a/drivers/amazon/net/ena/ena_ethtool.c b/drivers/amazon/net/ena/ena_ethtool.c
--- a/drivers/amazon/net/ena/ena_ethtool.c
+++ b/drivers/amazon/net/ena/ena_ethtool.c
@@ -648,8 +648,8 @@ static void ena_update_tx_rings_nonadaptive_intr_moderation(struct ena_adapter *
 	val = ena_com_get_nonadaptive_moderation_interval_tx(adapter->ena_dev);
 
 	for (i = 0; i < adapter->num_io_queues; i++) {
-		adapter->tx_ring[i].interrupt_interval_changed =
-			adapter->tx_ring[i].interrupt_interval != val;
+		adapter->tx_ring[i].interrupt_interval_changed |=
+			(adapter->tx_ring[i].interrupt_interval != val);
 		adapter->tx_ring[i].interrupt_interval = val;
 	}
 }
@@ -662,8 +662,8 @@ static void ena_update_rx_rings_nonadaptive_intr_moderation(struct ena_adapter *
 	val = ena_com_get_nonadaptive_moderation_interval_rx(adapter->ena_dev);
 
 	for (i = 0; i < adapter->num_io_queues; i++) {
-		adapter->rx_ring[i].interrupt_interval_changed =
-			adapter->rx_ring[i].interrupt_interval != val;
+		adapter->rx_ring[i].interrupt_interval_changed |=
+			(adapter->rx_ring[i].interrupt_interval != val);
 		adapter->rx_ring[i].interrupt_interval = val;
 	}
 }
diff --git a/drivers/amazon/net/ena/ena_netdev.c b/drivers/amazon/net/ena/ena_netdev.c
--- a/drivers/amazon/net/ena/ena_netdev.c
+++ b/drivers/amazon/net/ena/ena_netdev.c
@@ -1198,6 +1198,11 @@ static struct sk_buff *ena_rx_skb(struct ena_ring *rx_ring,
 		rx_ring->free_ids[*next_to_clean] = req_id;
 		*next_to_clean = ENA_RX_RING_IDX_ADD(*next_to_clean, descs,
 						     rx_ring->ring_size);
+
+		/* Don't reuse the RX page if we're on the wrong NUMA */
+		if (page_to_nid(rx_info->page) != numa_mem_id())
+			ena_free_rx_page(rx_ring, rx_info);
+
 		return skb;
 	}
 
@@ -2516,11 +2521,11 @@ int ena_up(struct ena_adapter *adapter)
 	 */
 	ena_init_napi_in_range(adapter, 0, io_queue_count);
 
-	/* Enabling DIM needs to happen before enabling IRQs since DIM
-	 * is run from napi routine
+	/* If the device stopped supporting interrupt moderation, need
+	 * to disable adaptive interrupt moderation.
 	 */
-	if (ena_com_interrupt_moderation_supported(adapter->ena_dev))
-		ena_com_enable_adaptive_moderation(adapter->ena_dev);
+	if (!ena_com_interrupt_moderation_supported(adapter->ena_dev))
+		ena_com_disable_adaptive_moderation(adapter->ena_dev);
 
 	rc = ena_request_io_irq(adapter);
 	if (rc)
@@ -4037,8 +4042,8 @@ static int check_for_rx_interrupt_queue(struct ena_adapter *adapter,
 	return 0;
 }
 
-enum ena_regs_reset_reason_types check_cdesc_in_tx_cq(struct ena_adapter *adapter,
-						      struct ena_ring *tx_ring)
+static enum ena_regs_reset_reason_types check_cdesc_in_tx_cq(struct ena_adapter *adapter,
+							     struct ena_ring *tx_ring)
 {
 	struct net_device *netdev = adapter->netdev;
 	u16 req_id;
@@ -4753,6 +4758,12 @@ static int ena_probe(struct pci_dev *pdev, const struct pci_device_id *ent)
 		goto err_device_destroy;
 	}
 
+	/* Enabling DIM needs to happen before enabling IRQs since DIM
+	 * is run from napi routine.
+	 */
+	if (ena_com_interrupt_moderation_supported(adapter->ena_dev))
+		ena_com_enable_adaptive_moderation(adapter->ena_dev);
+
 	ena_init_io_rings(adapter,
 			  0,
 			  adapter->xdp_num_queues +
diff --git a/drivers/amazon/net/ena/ena_netdev.h b/drivers/amazon/net/ena/ena_netdev.h
--- a/drivers/amazon/net/ena/ena_netdev.h
+++ b/drivers/amazon/net/ena/ena_netdev.h
@@ -32,7 +32,7 @@
 
 #define DRV_MODULE_GEN_MAJOR	2
 #define DRV_MODULE_GEN_MINOR	12
-#define DRV_MODULE_GEN_SUBMINOR 0
+#define DRV_MODULE_GEN_SUBMINOR	3
 
 #define DRV_MODULE_NAME		"ena"
 #ifndef DRV_MODULE_GENERATION
diff --git a/drivers/amazon/net/ena/kcompat.h b/drivers/amazon/net/ena/kcompat.h
--- a/drivers/amazon/net/ena/kcompat.h
+++ b/drivers/amazon/net/ena/kcompat.h
@@ -888,15 +888,7 @@ xdp_prepare_buff(struct xdp_buff *xdp, unsigned char *hard_start,
 #define ENA_XDP_XMIT_FREES_FAILED_DESCS_INTERNALLY
 #endif
 
-#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 15, 0) && \
-	!(LINUX_VERSION_CODE >= KERNEL_VERSION(5, 10, 188) && \
-	 LINUX_VERSION_CODE < KERNEL_VERSION(5, 11, 0)) && \
-	!(LINUX_VERSION_CODE >= KERNEL_VERSION(5, 4, 251) && \
-	 LINUX_VERSION_CODE < KERNEL_VERSION(5, 5, 0))) && \
-	!(defined(RHEL_RELEASE_CODE) && RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(8, 6)) && \
-	!(defined(SUSE_VERSION) && (SUSE_VERSION == 15 && SUSE_PATCHLEVEL >= 4)) && \
-	!(defined(SUSE_VERSION) && (SUSE_VERSION == 15 && SUSE_PATCHLEVEL == 3) && \
-	  ENA_KERNEL_VERSION_GTE(5, 3, 18, 150300, 59, 43))
+#ifndef ENA_HAVE_ETH_HW_ADDR_SET
 static inline void eth_hw_addr_set(struct net_device *dev, const u8 *addr)
 {
 	memcpy(dev->dev_addr, addr, ETH_ALEN);
