Date: Thu, 7 Aug 2025 10:42:06 +0000
Subject: Revert "netlink: Fix rmem check in netlink_broadcast_deliver()."

This reverts commit 37bc6baeb6696e3a1eca2d80c2dec92c276c9583.

Signed-off-by: Maximilian Heyne <mheyne@amazon.de>

diff --git a/net/netlink/af_netlink.c b/net/netlink/af_netlink.c
--- a/net/netlink/af_netlink.c
+++ b/net/netlink/af_netlink.c
@@ -1393,7 +1393,7 @@ static int netlink_broadcast_deliver(struct sock *sk, struct sk_buff *skb)
 	rmem = atomic_add_return(skb->truesize, &sk->sk_rmem_alloc);
 	rcvbuf = READ_ONCE(sk->sk_rcvbuf);
 
-	if ((rmem == skb->truesize || rmem <= rcvbuf) &&
+	if ((rmem != skb->truesize || rmem <= rcvbuf) &&
 	    !test_bit(NETLINK_S_CONGESTED, &nlk->state)) {
 		netlink_skb_set_owner_r(skb, sk);
 		__netlink_sendskb(sk, skb);
