Date: Thu, 22 Jan 2026 02:36:00 +0000
Subject: Revert "objtool: Fix weak symbol detection"

This reverts commit 57f91f25966724de26e8434387b2468e0a9a86fb.

Signed-off-by: Andrew Paniakin <apanyaki@amazon.com>

diff --git a/tools/objtool/elf.c b/tools/objtool/elf.c
--- a/tools/objtool/elf.c
+++ b/tools/objtool/elf.c
@@ -74,7 +74,7 @@ struct symbol_hole {
 };
 
 /*
- * Find the last symbol before @offset.
+ * Find !section symbol where @offset is after it.
  */
 static int symbol_hole_by_offset(const void *key, const struct rb_node *node)
 {
@@ -85,7 +85,8 @@ static int symbol_hole_by_offset(const void *key, const struct rb_node *node)
 		return -1;
 
 	if (sh->key >= s->offset + s->len) {
-		sh->sym = s;
+		if (s->type != STT_SECTION)
+			sh->sym = s;
 		return 1;
 	}
 
@@ -382,8 +383,7 @@ static void elf_add_symbol(struct elf *elf, struct symbol *sym)
 	sym->len = sym->sym.st_size;
 
 	__sym_for_each(iter, &sym->sec->symbol_tree, sym->offset, sym->offset) {
-		if (iter->offset == sym->offset && iter->type == sym->type &&
-		    iter->len == sym->len)
+		if (iter->offset == sym->offset && iter->type == sym->type)
 			iter->alias = sym;
 	}
 
