Date: Wed, 17 May 2023 18:19:59 +0000
Subject: AL2023 6.1 Update ena driver to 2.8.6g

Signed-off-by: Arthur Kiyanovski <akiyano@amazon.com>

diff --git a/drivers/amazon/net/ena/ena_netdev.c b/drivers/amazon/net/ena/ena_netdev.c
--- a/drivers/amazon/net/ena/ena_netdev.c
+++ b/drivers/amazon/net/ena/ena_netdev.c
@@ -693,16 +693,22 @@ static int ena_alloc_rx_buffer(struct ena_ring *rx_ring,
 	return 0;
 }
 
-static void ena_unmap_rx_buff(struct ena_ring *rx_ring,
-			      struct ena_rx_buffer *rx_info)
+static void ena_unmap_rx_buff_attrs(struct ena_ring *rx_ring,
+				    struct ena_rx_buffer *rx_info,
+				    unsigned long attrs)
 {
 	/* LPC pages are unmapped at cache destruction */
 	if (rx_info->is_lpc_page)
 		return;
 
-	dma_unmap_page(rx_ring->dev, rx_info->dma_addr,
-		       ENA_PAGE_SIZE,
-		       DMA_BIDIRECTIONAL);
+	ena_dma_unmap_page_attrs(rx_ring->dev, rx_info->dma_addr, ENA_PAGE_SIZE,
+				 DMA_BIDIRECTIONAL, attrs);
+}
+
+static void ena_unmap_rx_buff(struct ena_ring *rx_ring,
+			      struct ena_rx_buffer *rx_info)
+{
+	ena_unmap_rx_buff_attrs(rx_ring, rx_info, 0);
 }
 
 static void ena_free_rx_page(struct ena_ring *rx_ring,
@@ -880,7 +886,7 @@ static void ena_free_tx_bufs(struct ena_ring *tx_ring)
 
 		ena_unmap_tx_buff(tx_ring, tx_info);
 
-		napi_consume_skb(tx_info->skb, 0);
+		dev_kfree_skb_any(tx_info->skb);
 	}
 	netdev_tx_reset_queue(netdev_get_tx_queue(tx_ring->netdev,
 						  tx_ring->qid));
@@ -1012,7 +1018,7 @@ static int ena_clean_tx_irq(struct ena_ring *tx_ring, u32 budget)
 			  skb);
 
 		tx_bytes += tx_info->total_tx_size;
-		napi_consume_skb(skb, budget);
+		dev_kfree_skb(skb);
 		tx_pkts++;
 		total_done += tx_info->tx_descs;
 
@@ -1063,7 +1069,7 @@ static struct sk_buff *ena_alloc_skb(struct ena_ring *rx_ring, void *first_frag,
 	if (!first_frag)
 		skb = napi_alloc_skb(rx_ring->napi, len);
 	else
-		skb = ena_build_skb(first_frag, len);
+		skb = build_skb(first_frag, len);
 #else
 	if (!first_frag)
 		skb = napi_alloc_skb(rx_ring->napi, len);
@@ -1185,13 +1191,14 @@ static struct sk_buff *ena_rx_skb(struct ena_ring *rx_ring,
 	reuse_rx_buf_page = !is_xdp_loaded &&
 			    ena_try_rx_buf_page_reuse(rx_info, buf_len, len, pkt_offset);
 
+	dma_sync_single_for_cpu(rx_ring->dev,
+				pre_reuse_paddr + pkt_offset,
+				len,
+				DMA_FROM_DEVICE);
+
 	if (!reuse_rx_buf_page)
-		ena_unmap_rx_buff(rx_ring, rx_info);
-	else
-		dma_sync_single_for_cpu(rx_ring->dev,
-					pre_reuse_paddr + pkt_offset,
-					len,
-					DMA_FROM_DEVICE);
+		ena_unmap_rx_buff_attrs(rx_ring, rx_info, ENA_DMA_ATTR_SKIP_CPU_SYNC);
+
 
 	skb = ena_alloc_skb(rx_ring, buf_addr, buf_len);
 	if (unlikely(!skb))
@@ -1249,13 +1256,14 @@ static struct sk_buff *ena_rx_skb(struct ena_ring *rx_ring,
 		reuse_rx_buf_page = !is_xdp_loaded &&
 				    ena_try_rx_buf_page_reuse(rx_info, buf_len, len, pkt_offset);
 
+		dma_sync_single_for_cpu(rx_ring->dev,
+					pre_reuse_paddr + pkt_offset,
+					len,
+					DMA_FROM_DEVICE);
+
 		if (!reuse_rx_buf_page)
-			ena_unmap_rx_buff(rx_ring, rx_info);
-		else
-			dma_sync_single_for_cpu(rx_ring->dev,
-						pre_reuse_paddr + pkt_offset,
-						len,
-						DMA_FROM_DEVICE);
+			ena_unmap_rx_buff_attrs(rx_ring, rx_info, ENA_DMA_ATTR_SKIP_CPU_SYNC);
+
 
 		skb_add_rx_frag(skb, skb_shinfo(skb)->nr_frags, rx_info->page,
 				page_offset + buf_offset, len, buf_len);
@@ -3038,7 +3046,7 @@ static netdev_tx_t ena_start_xmit(struct sk_buff *skb, struct net_device *dev)
 	tx_info->skb = NULL;
 
 error_drop_packet:
-	napi_consume_skb(skb, 0);
+	dev_kfree_skb(skb);
 	return NETDEV_TX_OK;
 }
 
diff --git a/drivers/amazon/net/ena/ena_netdev.h b/drivers/amazon/net/ena/ena_netdev.h
--- a/drivers/amazon/net/ena/ena_netdev.h
+++ b/drivers/amazon/net/ena/ena_netdev.h
@@ -29,7 +29,7 @@
 
 #define DRV_MODULE_GEN_MAJOR	2
 #define DRV_MODULE_GEN_MINOR	8
-#define DRV_MODULE_GEN_SUBMINOR 3
+#define DRV_MODULE_GEN_SUBMINOR 6
 
 #define DRV_MODULE_NAME		"ena"
 #ifndef DRV_MODULE_GENERATION
diff --git a/drivers/amazon/net/ena/kcompat.h b/drivers/amazon/net/ena/kcompat.h
--- a/drivers/amazon/net/ena/kcompat.h
+++ b/drivers/amazon/net/ena/kcompat.h
@@ -107,6 +107,33 @@ Intel Corporation, 5200 N.E. Elam Young Parkway, Hillsboro, OR 97124-6497
 	LINUX_VERSION_CODE >= KERNEL_VERSION(3,10,0)
 #define ENA_BUSY_POLL_SUPPORT
 #endif
+
+/* Distribution kernel version comparison macros.
+ * Distribution kernel versioning format may be A.B.C-D.E.F and standard
+ * KERNEL_VERSION macro covers only the first 3 subversions.
+ * Using 20bit per subversion, as in some cases, subversion D may be a large
+ * number (6 digits).
+ */
+#define ENA_KERNEL_VERSION_16BIT(SV1, SV2, SV3) ((SV1 << 40) | (SV2 << 20) | (SV3))
+#define ENA_KERNEL_VERSION_MAJOR(SV1, SV2, SV3) ENA_KERNEL_VERSION_16BIT(SV1, SV2, SV3)
+#define ENA_KERNEL_VERSION_MINOR(SV1, SV2, SV3) ENA_KERNEL_VERSION_16BIT(SV1, SV2, SV3)
+
+#define ENA_KERNEL_VERSION_GTE(SV1, SV2, SV3, SV4, SV5, SV6) \
+	((ENA_KERNEL_VERSION_MAJOR(ENA_KERNEL_SUBVERSION_1, ENA_KERNEL_SUBVERSION_2, ENA_KERNEL_SUBVERSION_3) > \
+	  ENA_KERNEL_VERSION_MAJOR((SV1), (SV2), (SV3))) || \
+	 (ENA_KERNEL_VERSION_MAJOR(ENA_KERNEL_SUBVERSION_1, ENA_KERNEL_SUBVERSION_2, ENA_KERNEL_SUBVERSION_3) == \
+	  ENA_KERNEL_VERSION_MAJOR((SV1), (SV2), (SV3)) && \
+	  ENA_KERNEL_VERSION_MINOR(ENA_KERNEL_SUBVERSION_4, ENA_KERNEL_SUBVERSION_5, ENA_KERNEL_SUBVERSION_6) >= \
+	  ENA_KERNEL_VERSION_MINOR((SV4), (SV5), (SV6))))
+
+#define ENA_KERNEL_VERSION_LTE(SV1, SV2, SV3, SV4, SV5, SV6) \
+	((ENA_KERNEL_VERSION_MAJOR(ENA_KERNEL_SUBVERSION_1, ENA_KERNEL_SUBVERSION_2, ENA_KERNEL_SUBVERSION_3) < \
+	  ENA_KERNEL_VERSION_MAJOR((SV1), (SV2), (SV3))) || \
+	 (ENA_KERNEL_VERSION_MAJOR(ENA_KERNEL_SUBVERSION_1, ENA_KERNEL_SUBVERSION_2, ENA_KERNEL_SUBVERSION_3) == \
+	  ENA_KERNEL_VERSION_MAJOR((SV1), (SV2), (SV3)) && \
+	  ENA_KERNEL_VERSION_MINOR(ENA_KERNEL_SUBVERSION_4, ENA_KERNEL_SUBVERSION_5, ENA_KERNEL_SUBVERSION_6) <= \
+	  ENA_KERNEL_VERSION_MINOR((SV4), (SV5), (SV6))))
+
 /******************************************************************************/
 /************************** Ubuntu macros *************************************/
 /******************************************************************************/
@@ -177,7 +204,6 @@ Intel Corporation, 5200 N.E. Elam Young Parkway, Hillsboro, OR 97124-6497
 #define SUSE_VERSION 0
 #endif /* SUSE_VERSION */
 
-
 /******************************************************************************/
 /**************************** RHEL macros *************************************/
 /******************************************************************************/
@@ -822,17 +848,32 @@ static inline void netdev_rss_key_fill(void *buffer, size_t len)
 
 #if ((LINUX_VERSION_CODE < KERNEL_VERSION(4, 6 ,0)) && \
      !(RHEL_RELEASE_CODE && RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(7,3)))
-/* Linux versions 4.4.216 - 4.5 (non inclusive) back propagated page_ref_count
- * function from kernel 4.6. To make things more difficult, Ubuntu didn't add
- * these changes to its 4.4.* kernels
+/* Linux versions 4.4.216 - 4.5 (non inclusive) back propagated
+ * page_ref_count() from kernel 4.6.
+ * Ubuntu didn't add these changes to its 4.4.* kernels.
+ * UEK added this function in kernel 4.1.12-124.43.1
+ * Here is a figure that shows all of the cases:
+ * Legend:
+ * -------- page_ref_count() is present in the kernel
+ * ******** page_ref_count() is missing in the kernel
+ *
+ * Distro\Kernel      4.1.12-124.43.1   4.4.216    4.5    4.6
+ *                           |              |        |      |
+ * Upstrem kernel ***********|**************|--------|******|
+ *                           |              |        |      |
+ * Ubuntu         ***********|**************|********|******|
+ *                           |              |        |      |
+ * UEK            ***********|--------------|--------|------|
  */
-#if !(KERNEL_VERSION(4, 4 ,216) <= LINUX_VERSION_CODE && LINUX_VERSION_CODE < KERNEL_VERSION(4, 5 ,0)) ||\
-      defined(UBUNTU_VERSION_CODE)
+#if (defined(IS_UEK) && !ENA_KERNEL_VERSION_GTE(4, 1, 12, 124, 43, 1)) || \
+    (defined(ubuntu)) || \
+    (!defined(IS_UEK) && !defined(ubuntu) && \
+     !(KERNEL_VERSION(4, 4, 216) <= LINUX_VERSION_CODE && LINUX_VERSION_CODE < KERNEL_VERSION(4, 5, 0)))
 static inline int page_ref_count(struct page *page)
 {
 	return atomic_read(&page->_count);
 }
-#endif /* !(KERNEL_VERSION(4, 4 ,216) <= LINUX_VERSION_CODE && LINUX_VERSION_CODE < KERNEL_VERSION(4, 5 ,0)) */
+#endif /* (defined(IS_UEK) && !ENA_KERNEL_VERSION_GTE(4, 1, 12, 124, 43, 1)) ... */
 
 static inline void page_ref_inc(struct page *page)
 {
@@ -881,32 +922,16 @@ static inline int numa_mem_id(void)
 
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 5, 0)
 #define ENA_LINEAR_FRAG_SUPPORTED
-static __always_inline struct sk_buff*
-ena_build_skb(void *data, unsigned int frag_size)
-{
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 12, 0)
-	return napi_build_skb(data, frag_size);
-#else
-	return build_skb(data, frag_size);
-#endif
-}
-#endif
-
-#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 6, 0) && \
-	!(defined(RHEL_RELEASE_CODE) && RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(7, 3)) && \
-	!(defined(UBUNTU_VERSION_CODE) && UBUNTU_VERSION_CODE >= UBUNTU_VERSION(4, 2, 0, 42))
-static __always_inline
-void napi_consume_skb(struct sk_buff *skb, int budget)
-{
-	dev_kfree_skb_any(skb);
-}
 #endif
 
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 17, 0)
 #define ENA_NETDEV_LOGS_WITHOUT_RV
 #endif
 
-#if defined(ENA_XDP_SUPPORT) && LINUX_VERSION_CODE < KERNEL_VERSION(5, 12, 0)
+#if defined(ENA_XDP_SUPPORT) && \
+	(LINUX_VERSION_CODE < KERNEL_VERSION(5, 12, 0) && \
+	!(defined(SUSE_VERSION) && (SUSE_VERSION == 15 && SUSE_PATCHLEVEL == 3) && \
+	 ENA_KERNEL_VERSION_GTE(5, 3, 18, 150300, 59, 49)))
 static __always_inline void
 xdp_init_buff(struct xdp_buff *xdp, u32 frame_sz, struct xdp_rxq_info *rxq)
 {
@@ -928,7 +953,7 @@ xdp_prepare_buff(struct xdp_buff *xdp, unsigned char *hard_start,
 	xdp->data_meta = meta_valid ? data : data + 1;
 }
 
-#endif /* defined(ENA_XDP_SUPPORT) && LINUX_VERSION_CODE <= KERNEL_VERSION(5, 12, 0) */
+#endif /* defined(ENA_XDP_SUPPORT) && (LINUX_VERSION_CODE <= KERNEL_VERSION(5, 12, 0) && !SUSE_VERSION(...)) */
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(5, 13, 0)
 #define ethtool_sprintf(data, fmt, args...)			\
@@ -944,7 +969,9 @@ xdp_prepare_buff(struct xdp_buff *xdp, unsigned char *hard_start,
 
 #if LINUX_VERSION_CODE < KERNEL_VERSION(5, 15, 0) && \
 	!(defined(RHEL_RELEASE_CODE) && RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(8, 6)) && \
-	!(defined(SUSE_VERSION) && (SUSE_VERSION == 15 && SUSE_PATCHLEVEL >= 4))
+	!(defined(SUSE_VERSION) && (SUSE_VERSION == 15 && SUSE_PATCHLEVEL >= 4)) && \
+	!(defined(SUSE_VERSION) && (SUSE_VERSION == 15 && SUSE_PATCHLEVEL == 3) && \
+	  ENA_KERNEL_VERSION_GTE(5, 3, 18, 150300, 59, 43))
 static inline void eth_hw_addr_set(struct net_device *dev, const u8 *addr)
 {
 	memcpy(dev->dev_addr, addr, ETH_ALEN);
@@ -1065,7 +1092,8 @@ static inline void ena_netif_napi_add(struct net_device *dev,
 				      struct napi_struct *napi,
 				      int (*poll)(struct napi_struct *, int))
 {
-#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 1, 0)
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 1, 0)) && \
+	!(RHEL_RELEASE_CODE && (RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(9, 2)))
 #ifndef NAPI_POLL_WEIGHT
 #define NAPI_POLL_WEIGHT 64
 #endif
@@ -1075,4 +1103,52 @@ static inline void ena_netif_napi_add(struct net_device *dev,
 #endif /* LINUX_VERSION_CODE < KERNEL_VERSION(6, 1, 0) */
 }
 
+#if (RHEL_RELEASE_CODE && (RHEL_RELEASE_CODE <= RHEL_RELEASE_VERSION(7, 4))) || \
+    (defined(UBUNTU_VERSION_CODE) && UBUNTU_VERSION_CODE < UBUNTU_VERSION(4, 5, 0, 0)) || \
+    (defined(IS_UEK) && !ENA_KERNEL_VERSION_GTE(4, 1, 12, 105, 0, 0))
+static inline void dma_unmap_page_attrs(struct device *dev,
+					dma_addr_t addr, size_t size,
+					enum dma_data_direction dir,
+					struct dma_attrs *attrs)
+{
+	struct dma_map_ops *ops = get_dma_ops(dev);
+
+	BUG_ON(!valid_dma_direction(dir));
+	if (ops->unmap_page)
+		ops->unmap_page(dev, addr, size, dir, attrs);
+	debug_dma_unmap_page(dev, addr, size, dir, false);
+}
+#endif /* RHEL_RELEASE_CODE && (RHEL_RELEASE_CODE <= RHEL_RELEASE_VERSION(7, 4)) */
+
+#if (RHEL_RELEASE_CODE && (RHEL_RELEASE_CODE <= RHEL_RELEASE_VERSION(7, 9)) && \
+     (RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(7, 0)) && \
+     (LINUX_VERSION_CODE != KERNEL_VERSION(4, 14, 0))) || \
+    (defined(UBUNTU_VERSION_CODE) && UBUNTU_VERSION_CODE < UBUNTU_VERSION(4, 5, 0, 0)) || \
+    (defined(IS_UEK) && LINUX_VERSION_CODE < KERNEL_VERSION(4, 1, 13))
+#define ENA_DMA_ATTR_SKIP_CPU_SYNC (1 << DMA_ATTR_SKIP_CPU_SYNC)
+#elif (RHEL_RELEASE_CODE && (RHEL_RELEASE_CODE <= RHEL_RELEASE_VERSION(6, 10)))
+#define	ENA_DMA_ATTR_SKIP_CPU_SYNC 0
+#else
+#define ENA_DMA_ATTR_SKIP_CPU_SYNC DMA_ATTR_SKIP_CPU_SYNC
+#endif
+
+static inline void ena_dma_unmap_page_attrs(struct device *dev,
+					    dma_addr_t addr, size_t size,
+					    enum dma_data_direction dir,
+					    unsigned long attrs)
+{
+#if (RHEL_RELEASE_CODE && (RHEL_RELEASE_CODE <= RHEL_RELEASE_VERSION(7, 9)) && \
+     (LINUX_VERSION_CODE != KERNEL_VERSION(4, 14, 0))) || \
+    (defined(UBUNTU_VERSION_CODE) && UBUNTU_VERSION_CODE < UBUNTU_VERSION(4, 5, 0, 0)) || \
+    (defined(IS_UEK) && LINUX_VERSION_CODE < KERNEL_VERSION(4, 1, 13))
+	struct dma_attrs dma_attrs;
+
+	init_dma_attrs(&dma_attrs);
+	dma_attrs.flags[0] = attrs;
+	dma_unmap_page_attrs(dev, addr, size, dir, &dma_attrs);
+#else
+	dma_unmap_page_attrs(dev, addr, size, dir, attrs);
+#endif
+}
+
 #endif /* _KCOMPAT_H_ */
