Date: Wed, 7 Feb 2024 19:39:05 +0000
Subject: AL2023-6.1-Update-ena-driver-to-2.11.1g

Also some minor misalignment in kcompat.h vs 2.11.0
to void a diff.

Signed-off-by: David Arinzon <darinzon@amazon.com>

diff --git a/drivers/amazon/net/ena/Makefile b/drivers/amazon/net/ena/Makefile
--- a/drivers/amazon/net/ena/Makefile
+++ b/drivers/amazon/net/ena/Makefile
@@ -1,7 +1,7 @@
 #
 # Makefile for the Elastic Network Adapter (ENA) device drivers.
 # ENA Source is: https://github.com/amzn/amzn-drivers.
-# Current ENA source is based on ena_linux_2.11.0 tag.
+# Current ENA source is based on ena_linux_2.11.1 tag.
 #
 
 obj-$(CONFIG_AMAZON_ENA_ETHERNET) += ena.o
diff --git a/drivers/amazon/net/ena/ena_netdev.c b/drivers/amazon/net/ena/ena_netdev.c
--- a/drivers/amazon/net/ena/ena_netdev.c
+++ b/drivers/amazon/net/ena/ena_netdev.c
@@ -4914,6 +4914,8 @@ static void __ena_shutoff(struct pci_dev *pdev, bool shutdown)
 	adapter->reset_reason = ENA_REGS_RESET_SHUTDOWN;
 	ena_destroy_device(adapter, true);
 
+	ena_phc_free(adapter);
+
 	if (shutdown) {
 		netif_device_detach(netdev);
 		dev_close(netdev);
@@ -4932,8 +4934,6 @@ static void __ena_shutoff(struct pci_dev *pdev, bool shutdown)
 
 	ena_com_delete_customer_metrics_buffer(ena_dev);
 
-	ena_phc_free(adapter);
-
 	ena_release_bars(ena_dev, pdev);
 
 	pci_disable_device(pdev);
diff --git a/drivers/amazon/net/ena/ena_netdev.h b/drivers/amazon/net/ena/ena_netdev.h
--- a/drivers/amazon/net/ena/ena_netdev.h
+++ b/drivers/amazon/net/ena/ena_netdev.h
@@ -32,7 +32,7 @@
 
 #define DRV_MODULE_GEN_MAJOR	2
 #define DRV_MODULE_GEN_MINOR	11
-#define DRV_MODULE_GEN_SUBMINOR 0
+#define DRV_MODULE_GEN_SUBMINOR 1
 
 #define DRV_MODULE_NAME		"ena"
 #ifndef DRV_MODULE_GENERATION
diff --git a/drivers/amazon/net/ena/kcompat.h b/drivers/amazon/net/ena/kcompat.h
--- a/drivers/amazon/net/ena/kcompat.h
+++ b/drivers/amazon/net/ena/kcompat.h
@@ -532,7 +532,7 @@ static inline bool ena_u64_stats_fetch_retry(const struct u64_stats_sync *syncp,
 					     unsigned int start)
 {
 #if LINUX_VERSION_CODE < KERNEL_VERSION(6, 1, 0) && \
-    !(RHEL_RELEASE_CODE && RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(9, 3))
+    !(RHEL_RELEASE_CODE && RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(9,3))
 	return u64_stats_fetch_retry_irq(syncp, start);
 #else
 	return u64_stats_fetch_retry(syncp, start);
@@ -542,7 +542,7 @@ static inline bool ena_u64_stats_fetch_retry(const struct u64_stats_sync *syncp,
 static inline unsigned int ena_u64_stats_fetch_begin(const struct u64_stats_sync *syncp)
 {
 #if LINUX_VERSION_CODE < KERNEL_VERSION(6, 1, 0) && \
-    !(RHEL_RELEASE_CODE && RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(9, 3))
+    !(RHEL_RELEASE_CODE && RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(9,3))
 	return u64_stats_fetch_begin_irq(syncp);
 #else
 	return u64_stats_fetch_begin(syncp);
